### PR TITLE
feat: basic symmetric predicates

### PR DESCRIPTION
**Pull request by liangyiliang**
# Description

Related issue/PR: #744 

The Penrose languages now support basic "symmetric" predicates. If a predicate `MyPred` is declared `symmetric`, then within the Style program, `MyPred(a, b)` also matches `MyPred(b, a)` automatically. We currently only support binary predicates.

# Implementation strategy and design decisions

* We modified the Domain parser to parse `symmetric predicate` correctly.
* We modified the Domain compiler to enforce some conditions for symmetric predicates: all symmetric predicates can only have two arguments, and they must have the same type.
* We modified the Style compiler to take symmetric predicates into consideration when matching predicates.
* We added an example that corresponds to the Water-bond example in issue #744.

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new ESLint warnings
- [x] I have reviewed any generated changes to the `diagrams/` folder

### CODE REVIEW

File: packages/core/src/parser/Domain.ne [hunk 1]
```diff
@@ -82,11 +83,15 @@ type_decl -> "type" __ identifier (_ "(" _ type_params _ ")"):? (_ "<:" _ sepBy1
   }
 %}
 
-predicate -> "predicate" __ identifier type_params_list args_list {%
-  ([kw, , name, params, args]): PredicateDecl<C> => ({
+# This now works with symmetric predicate declarations.
+predicate -> ("symmetric" __):? "predicate" __ identifier type_params_list args_list {%
+  ([sym, kw, , name, params, args]): PredicateDecl<C> => ({
     ...nodeData,
     ...rangeFrom([rangeOf(kw), ...args, ...params]),
```
Comment (wodeni) [inline packages/core/src/parser/Domain.ne:90]: Minor point: if qualifiers like `symmetric` come before `predicate`, the starting source position will need to change. Now that I think about it, I kind of like `predicate symmetric` a bit more? Not sure if the concrete syntax was discussed at all

Comment (wodeni) [reply packages/core/src/parser/Domain.ne:90]: Yep that seems right!

Comment (liangyiliang) [reply packages/core/src/parser/Domain.ne]: I just pushed some new code if you want to take a look!

**Comment by joshsunshine** [comment]
It looks great overall. Can you edit the style program to make the Hydrogens closer to the Oxygen molecules in the example?

**Comment by liangyiliang** [comment]
Sure I can do that - hold on

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Comment by liangyiliang** [comment]
Just fixed the diagram ...

File: packages/examples/src/symmetric-bond/symmetric-bond.sty [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    v1 = o.bond1.end - o.bond1.start
+    v2 = o.bond2.end - o.bond2.start
+
+    o.layering1 = h1.icon above o.bond1
+    o.layering2 = h2.icon above o.bond2
+    o.layering3 = o.icon above o.bond1
+    
+    ensure equal(norm(v1), norm(v2))
+    ensure equal(angleBetween(v1, v2), toRadians(104.5))
+    ensure equal(norm(v1), 100)
```
Comment (joshsunshine) [inline packages/examples/src/symmetric-bond/symmetric-bond.sty:54]: This seems reasonable and the diagram looks better.

**Pull request merged**